clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -4793,6 +4803,155 @@ void CodeGenModule::addGlobalIntelFPGAAnnotation(const VarDecl *VD,
   }
 }
 
+static Optional<std::string>
+getValidAttributeNameAsString(const Expr *NameE, const ASTContext &Context) {
+  if (const auto *NameLiteral = dyn_cast<StringLiteral>(NameE))
+    return NameLiteral->getString().str();
+
+  const auto *NameCE = dyn_cast<ConstantExpr>(NameE);
+  APValue NameLValue;
+  if (!NameCE || !NameCE->isCXX11ConstantExpr(Context, &NameLValue) ||
+      !NameLValue.isLValue())
+    return None;
+
+  const auto *NameValExpr = NameLValue.getLValueBase().dyn_cast<const Expr *>();
+  if (NameValExpr)
+    return getValidAttributeNameAsString(NameValExpr, Context);
+
+  const auto *NameValDecl =
+      NameLValue.getLValueBase().dyn_cast<const ValueDecl *>();
+  if (!NameValDecl)
+    return None;
+
+  const auto *NameVarDecl = dyn_cast<const VarDecl>(NameValDecl);
+  if (!NameVarDecl)
+    return None;
+
+  return getValidAttributeNameAsString(NameVarDecl->getInit(), Context);
+}
+
+static Optional<std::string>
+getValidAttributeValueAsString(const APValue Value, const ASTContext &Context,
+                               QualType ValueQType) {
+  assert(!Value.isLValue());
+  if (ValueQType->isCharType()) {
+    char C = static_cast<char>(Value.getInt().getExtValue());
+    return std::string(&C, 1);
+  }
+  if (ValueQType->isBooleanType())
+    return std::string(Value.getInt().getExtValue() ? "true" : "false");
+  if (ValueQType->isIntegerType() || ValueQType->isFloatingType() ||
+      ValueQType->isEnumeralType())
+    return Value.getAsString(Context, ValueQType);
+  return None;
+}
+
+static Optional<std::string>
+getValidAttributeValueAsString(const Expr *ValueE, const ASTContext &Context) {
+  if (ValueE->getType()->isNullPtrType())
+    return std::string("");
```

Comment:
This seems suspicious but may be fine -- I would expect that an explicit `nullptr` would map to something other than an empty string though.